### PR TITLE
Added kafka message versioning with ZooKeeper

### PR DESCRIPTION
**Pull request by niksv**
This PR adds kafka Header to all Kafka messages in Kilda. This is needed to isolate components by version.
If a component read Kafka message without version or with different version - this message will be skipped. 

Each component gets current version from ZooKeeper during start up. If ZooKeeper has no version for component default version will be used.
At this moment all components will use default version. When Zero Downtime feature will be finished Kilda will use some other version.

### CODE REVIEW

**Review by nikitamarchenko — CHANGES_REQUESTED** [review]

File: src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/kafka/versioning/VersioningConsumerInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@Slf4j
+public class VersioningConsumerInterceptor<K, V> extends VersioningInterceptorBase
+        implements ConsumerInterceptor<K, V>, BuildVersionObserver {
+
+    public VersioningConsumerInterceptor() {
+        log.info("Initializing VersioningConsumerInterceptor");
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/kafka/versioning/VersioningConsumerInterceptor.java:48]: fixed

File: src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/kafka/versioning/VersioningProducerInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Map;
+
+@Slf4j
+public class VersioningProducerInterceptor<K, V> extends VersioningInterceptorBase
+        implements ProducerInterceptor<K, V>, BuildVersionObserver {
+
+    public VersioningProducerInterceptor() {
+        log.info("Initializing VersioningProducerInterceptor");
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/kafka/versioning/VersioningProducerInterceptor.java:41]: fixed

File: src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/kafka/versioning/VersioningProducerInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            ZkWatchDog watchDog = ZkWatchDog.builder()
+                    .id(runId)
+                    .serviceName(componentName)
+                    .connectionString(connectionString).build();
+
+            watchDog.subscribe(this);
+        } catch (IOException e) {
+            log.error("Component {} with id {} can't connect to ZooKeeper with connection string: {}, received: {}",
+                    componentName, runId, connectionString, e.getMessage());
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/kafka/versioning/VersioningProducerInterceptor.java:96]: fixed

File: src-java/blue-green/src/main/java/org/openkilda/bluegreen/ZkClient.java [hunk 4]
```diff
@@ -73,10 +73,14 @@ protected ZooKeeper getZk() throws IOException {
     }
 
     protected void ensureZNode(String... path) throws KeeperException, InterruptedException {
+        ensureZNode("".getBytes(), path);
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/blue-green/src/main/java/org/openkilda/bluegreen/ZkClient.java:76]: fixed

File: src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/kafka/versioning/VersioningProducerInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                log.warn("Messaging version is not set for component {} with id {}. Skip record {}",
+                        componentName, runId, record);
+                versionIsNotSetTimestamp = Instant.now();
+            }
+            // We can't return null record. We will return record without version.
+            // Such records wouldn't be read by Kafka Consumer with VersioningConsumerInterceptor
+            return record;
+        }
+
+        record.headers().add(MESSAGE_VERSION_HEADER, version.getBytes());
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/kafka/versioning/VersioningProducerInterceptor.java:65]: fixed

File: src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/kafka/versioning/VersioningConsumerInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    @Override
+    public void onCommit(Map<TopicPartition, OffsetAndMetadata> offsets) {
+        // nothing to do here
+    }
+
+    @Override
+    public void close() {
+        // nothing to do here
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/kafka/versioning/VersioningConsumerInterceptor.java:133]: fixed

File: src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/kafka/versioning/VersioningConsumerInterceptor.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+        try {
+            ZkWatchDog watchDog = ZkWatchDog.builder()
+                    .id(runId)
+                    .serviceName(componentName)
+                    .connectionString(connectionString).build();
+
+            watchDog.subscribe(this);
+        } catch (IOException e) {
+            log.error("Component {} with id {} can't connect to ZooKeeper with connection string: {}, received: {}",
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/kafka/versioning/VersioningConsumerInterceptor.java]: fixed

File: src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/kafka/versioning/VersioningProducerInterceptor.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+        try {
+            ZkWatchDog watchDog = ZkWatchDog.builder()
+                    .id(runId)
+                    .serviceName(componentName)
+                    .connectionString(connectionString).build();
+
+            watchDog.subscribe(this);
+        } catch (IOException e) {
+            log.error("Component {} with id {} can't connect to ZooKeeper with connection string: {}, received: {}",
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/kafka/versioning/VersioningProducerInterceptor.java]: fixed

File: src-java/base-topology/base-storm-topology/src/test/resources/topology.properties [hunk 9]
```diff
@@ -6,7 +6,7 @@ parallelism = 1
 flow.hs.parallelism = 4
 workers = 1
 
-zookeeper.hosts=localhost:2181
+zookeeper.hosts=localhost:2181/kilda
```
[the remark this answers is not in the document]
Comment (niksv) [reply src-java/base-topology/base-storm-topology/src/test/resources/topology.properties]: fixed

File: src-java/server42/server42-stats/src/main/java/org/openkilda/server42/stats/kafka/KafkaConfig.java [hunk 10]
```diff
@@ -15,16 +15,61 @@
...
+    @Value("${zookeeper.connect_string}")
+    private String zookeeperConnectString;
+
     @Bean
     public NewTopic toStorm(@Value("${openkilda.server42.stats.kafka.topic.flowrtt.to_storm}") String name) {
         return TopicBuilder.name(name).build();
     }
+
+    @Bean
+    public ConsumerFactory<?, ?> kafkaConsumerFactory(KafkaProperties properties) {
```
Comment (nikitamarchenko) [inline src-java/server42/server42-stats/src/main/java/org/openkilda/server42/stats/kafka/KafkaConfig.java:57]: move it config files

File: src-java/server42/server42-stats/src/main/java/org/openkilda/server42/stats/kafka/KafkaConfig.java [hunk 11]
```diff
@@ -15,16 +15,61 @@
...
+        Map<String, Object> updatedProperties = properties.buildConsumerProperties();
+        updatedProperties.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, VersioningConsumerInterceptor.class.getName());
+        updatedProperties.put(CONSUMER_COMPONENT_NAME_PROPERTY, SERVER_42_STATS_COMPONENT_NAME);
+        updatedProperties.put(CONSUMER_RUN_ID_PROPERTY, SERVER_42_RUN_ID);
+        updatedProperties.put(CONSUMER_ZOOKEEPER_CONNECTION_STRING_PROPERTY, zookeeperConnectString);
+        return new DefaultKafkaConsumerFactory<>(updatedProperties);
+    }
+
+    @Bean
+    public ProducerFactory<?, ?> kafkaProducerFactory(KafkaProperties properties) {
```
Comment (nikitamarchenko) [inline src-java/server42/server42-stats/src/main/java/org/openkilda/server42/stats/kafka/KafkaConfig.java:67]: move it config files

File: src-java/server42/server42-control/src/main/java/org/openkilda/server42/control/kafka/KafkaConfig.java [hunk 12]
```diff
@@ -49,4 +73,24 @@ public NewTopic fromStorm(@Value("${openkilda.server42.control.kafka.topic.from_
         factory.setRecordFilterStrategy(kafkaRecordFilter);
         return factory;
     }
+
+    @Bean
+    public ConsumerFactory<?, ?> kafkaConsumerFactory(KafkaProperties properties) {
```
Comment (nikitamarchenko) [inline src-java/server42/server42-control/src/main/java/org/openkilda/server42/control/kafka/KafkaConfig.java:78]: move it config files

File: src-java/server42/server42-control/src/main/java/org/openkilda/server42/control/kafka/KafkaConfig.java [hunk 13]
```diff
@@ -49,4 +73,24 @@ public NewTopic fromStorm(@Value("${openkilda.server42.control.kafka.topic.from_
...
+        Map<String, Object> updatedProperties = properties.buildConsumerProperties();
+        updatedProperties.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, VersioningConsumerInterceptor.class.getName());
+        updatedProperties.put(CONSUMER_COMPONENT_NAME_PROPERTY, SERVER_42_CONTROL_COMPONENT_NAME);
+        updatedProperties.put(CONSUMER_RUN_ID_PROPERTY, SERVER_42_RUN_ID);
+        updatedProperties.put(CONSUMER_ZOOKEEPER_CONNECTION_STRING_PROPERTY, zookeeperConnectString);
+        return new DefaultKafkaConsumerFactory<>(updatedProperties);
+    }
+
+    @Bean
+    public ProducerFactory<?, ?> kafkaProducerFactory(KafkaProperties properties) {
```
Comment (nikitamarchenko) [inline src-java/server42/server42-control/src/main/java/org/openkilda/server42/control/kafka/KafkaConfig.java:88]: move it config files

**Review by nikitamarchenko — CHANGES_REQUESTED** [review]

File: src-java/server42/server42-stats/src/main/java/org/openkilda/server42/stats/kafka/KafkaConfig.java (hunk 10 above)
Comment (niksv) [reply src-java/server42/server42-stats/src/main/java/org/openkilda/server42/stats/kafka/KafkaConfig.java]: fixed

File: src-java/server42/server42-control/src/main/java/org/openkilda/server42/control/kafka/KafkaConfig.java (hunk 12 above)
Comment (niksv) [reply src-java/server42/server42-control/src/main/java/org/openkilda/server42/control/kafka/KafkaConfig.java]: fixed

File: src-java/server42/server42-stats/src/main/java/org/openkilda/server42/stats/kafka/KafkaConfig.java (hunk 11 above)
Comment (niksv) [reply src-java/server42/server42-stats/src/main/java/org/openkilda/server42/stats/kafka/KafkaConfig.java]: fixed

File: src-java/server42/server42-control/src/main/java/org/openkilda/server42/control/kafka/KafkaConfig.java (hunk 13 above)
Comment (niksv) [reply src-java/server42/server42-control/src/main/java/org/openkilda/server42/control/kafka/KafkaConfig.java]: fixed

**Review by nikitamarchenko — APPROVED** [review]

**Review by dpoltavets — APPROVED** [review]

**Pull request merged**